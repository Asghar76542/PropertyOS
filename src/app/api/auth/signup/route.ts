import { NextRequest, NextResponse } from 'next/server'
import bcrypt from 'bcryptjs'
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

export async function POST(request: NextRequest) {
  try {
    const body = await request.json()
    const {
      email,
      password,
      firstName,
      lastName,
      phone,
      userType,
      // Landlord fields
      companyName,
      businessAddress,
      businessCity,
      businessPostcode,
      // Tenant fields
      dateOfBirth,
      currentAddress,
      city,
      postcode,
      employmentStatus,
      monthlyIncome
    } = body

    // Validate required fields
    if (!email || !password || !firstName || !lastName || !phone || !userType) {
      return NextResponse.json(
        { message: 'Missing required fields' },
        { status: 400 }
      )
    }

    // Check if user already exists
    const existingUser = await prisma.user.findUnique({
      where: { email }
    })

    if (existingUser) {
      return NextResponse.json(
        { message: 'User with this email already exists' },
        { status: 400 }
      )
    }

    // Hash password
    const hashedPassword = await bcrypt.hash(password, 12)

    // Create user with role-specific data
    const userData = {
      email,
      password: hashedPassword,
      firstName,
      lastName,
      phone,
      role: userType.toUpperCase() as 'LANDLORD' | 'TENANT',
      createdAt: new Date(),
      updatedAt: new Date()
    }

    const user = await prisma.user.create({
      data: userData
    })

    // Create role-specific profile
    if (userType === 'landlord') {
      await prisma.landlordProfile.create({
        data: {
          userId: user.id,
          companyName: companyName || null,
          businessAddress: businessAddress || '',
          businessCity: businessCity || '',
          businessPostcode: businessPostcode || '',
          createdAt: new Date(),
          updatedAt: new Date()
        }
      })
    } else if (userType === 'tenant') {
      await prisma.tenantProfile.create({
        data: {
          userId: user.id,
          dateOfBirth: dateOfBirth ? new Date(dateOfBirth) : null,
          currentAddress: currentAddress || '',
          city: city || '',
          postcode: postcode || '',
          employmentStatus: employmentStatus || null,
          monthlyIncome: monthlyIncome || null,
          createdAt: new Date(),
          updatedAt: new Date()
        }
      })
    }

    // Return success (don't send password)
    const { password: _, ...userWithoutPassword } = user

    return NextResponse.json({
      message: 'Account created successfully',
      user: userWithoutPassword
    }, { status: 201 })

  } catch (error) {
    console.error('Signup error:', error)
    return NextResponse.json(
      { message: 'Internal server error' },
      { status: 500 }
    )
  } finally {
    await prisma.$disconnect()
  }
}
