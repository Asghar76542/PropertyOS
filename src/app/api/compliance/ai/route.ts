import { NextRequest, NextResponse } from 'next/server'
import ZAI from 'z-ai-web-dev-sdk'

export async function POST(request: NextRequest) {
  try {
    const { query, propertyContext } = await request.json()

    if (!query) {
      return NextResponse.json(
        { error: 'Query is required' },
        { status: 400 }
      )
    }

    // Initialize ZAI SDK
    const zai = await ZAI.create()

    // Create a comprehensive prompt for property compliance
    const systemPrompt = `You are an expert UK property compliance advisor with deep knowledge of:
- Gas Safety Regulations
- Energy Performance Certificates (EPC)
- Electrical Safety Standards
- Fire Safety Regulations
- HMO Licensing
- Tenant Deposit Protection
- Right to Rent Checks
- Building Regulations
- Housing Health and Safety Rating System (HHSRS)
- Legionella Risk Assessment
- Asbestos Management
- Carbon Monoxide Alarm Requirements

Provide accurate, up-to-date compliance guidance for UK landlords. Include specific regulatory references where applicable and give practical, actionable advice. Consider the property context provided and tailor your response accordingly.`

    const userPrompt = `Compliance Query: ${query}

${propertyContext ? `Property Context: ${propertyContext}` : ''}

Please provide comprehensive compliance guidance including:
1. Regulatory requirements
2. Deadlines and timelines
3. Potential penalties for non-compliance
4. Recommended actions
5. Best practices`

    // Get AI completion
    const completion = await zai.chat.completions.create({
      messages: [
        {
          role: 'system',
          content: systemPrompt
        },
        {
          role: 'user',
          content: userPrompt
        }
      ],
      temperature: 0.3, // Lower temperature for more factual responses
      max_tokens: 1000
    })

    const aiResponse = completion.choices[0]?.message?.content

    if (!aiResponse) {
      return NextResponse.json(
        { error: 'Failed to generate AI response' },
        { status: 500 }
      )
    }

    return NextResponse.json({
      response: aiResponse,
      timestamp: new Date().toISOString()
    })

  } catch (error) {
    console.error('Error in AI compliance advisor:', error)
    return NextResponse.json(
      { error: 'Failed to process AI request' },
      { status: 500 }
    )
  }
}