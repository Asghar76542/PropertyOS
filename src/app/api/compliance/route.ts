import { NextResponse } from 'next/server'
import { db } from '@/lib/db'
import { getServerSession } from 'next-auth'
import { authOptions } from '@/lib/auth'
import { z } from 'zod'
import { logger } from '@/lib/logger'
import { AppError, ErrorCode, generateRequestId } from '@/lib/errors'
import { createAuditLog } from '@/lib/audit'

const complianceRecordSchema = z.object({
  title: z.string().min(1, "Title is required"),
  description: z.string().min(1, "Description is required"),
  requirement: z.string().min(1, "Requirement is required"),
  dueDate: z.string().datetime(),
  propertyId: z.string().min(1, "Property ID is required"),
})

export async function GET(request: Request) {
  const requestId = generateRequestId();
  let session;
  try {
    session = await getServerSession(authOptions)
    if (!session?.user?.id) {
      throw new AppError('Unauthorized', 401, ErrorCode.UNAUTHORIZED);
    }

    if (session.user.role !== 'LANDLORD' && session.user.role !== 'ADMIN') {
      throw new AppError('Insufficient permissions', 403, ErrorCode.INSUFFICIENT_PERMISSIONS);
    }

    const landlordId = session.user.id;

    const complianceRecords = await db.complianceRecord.findMany({
        where: { property: { landlordId: landlordId } },
        include: {
            property: true,
            document: true,
        }
    });

    const calculateComplianceRate = (records: any[], category: string) => {
        const categoryRecords = records.filter(r => r.title.includes(category));
        if (categoryRecords.length === 0) return 0;
        const compliantRecords = categoryRecords.filter(r => r.status === 'COMPLIANT').length;
        return (compliantRecords / categoryRecords.length) * 100;
    }

    const complianceStatus = {
        overall: complianceRecords.length > 0 ? (complianceRecords.filter(r => r.status === 'COMPLIANT').length / complianceRecords.length) * 100 : 0,
        gasSafety: calculateComplianceRate(complianceRecords, 'Gas Safety'),
        epc: calculateComplianceRate(complianceRecords, 'EPC'),
        electrical: calculateComplianceRate(complianceRecords, 'Electrical'),
        fireSafety: calculateComplianceRate(complianceRecords, 'Fire Safety'),
        hmoLicense: calculateComplianceRate(complianceRecords, 'HMO')
    };

    return NextResponse.json({ 
        complianceRecords,
        complianceStatus
    });

  } catch (error) {
    logger.error('Error fetching compliance data:', { error, requestId, userId: session?.user?.id })
    
    if (error instanceof AppError) {
      return NextResponse.json(
        { 
          error: error.message,
          code: error.code,
          details: error.details 
        }, 
        { status: error.statusCode }
      )
    }
    
    return NextResponse.json(
      { 
        error: 'Failed to fetch compliance data',
        code: ErrorCode.INTERNAL_ERROR,
        requestId: requestId
      }, 
      { status: 500 }
    )
  }
}

export async function POST(request: Request) {
  const requestId = generateRequestId();
  let session;
  let validatedData;
  try {
    session = await getServerSession(authOptions)
    if (!session?.user?.id) {
      throw new AppError('Unauthorized', 401, ErrorCode.UNAUTHORIZED);
    }

    if (session.user.role !== 'LANDLORD' && session.user.role !== 'ADMIN') {
      throw new AppError('Insufficient permissions', 403, ErrorCode.INSUFFICIENT_PERMISSIONS);
    }

    const userId = session.user.id;
    const body = await request.json();

    validatedData = complianceRecordSchema.parse(body);

    const complianceRecord = await db.complianceRecord.create({
      data: {
        title: validatedData.title,
        description: validatedData.description,
        requirement: validatedData.requirement,
        dueDate: new Date(validatedData.dueDate),
        propertyId: validatedData.propertyId,
        userId: userId,
        status: 'PENDING'
      }
    })

    await createAuditLog({
        userId: session.user.id,
        action: 'create_compliance_record',
        resource: 'compliance_record',
        resourceId: complianceRecord.id,
        changes: { after: complianceRecord },
        result: 'success',
    });

    return NextResponse.json({ complianceRecord }, { status: 201 });
  } catch (error) {
    const errorDetails = error instanceof Error ? error.message : 'Unknown error';
    await createAuditLog({
        userId: session?.user?.id || 'unknown',
        action: 'create_compliance_record',
        resource: 'compliance_record',
        resourceId: 'n/a',
        changes: { before: validatedData || {} },
        result: 'failure',
        errorDetails,
    });

    logger.error('Error creating compliance record:', { error, requestId, userId: session?.user?.id })

    if (error instanceof z.ZodError) {
        return NextResponse.json(
            {
                error: 'Invalid input',
                code: ErrorCode.VALIDATION_FAILED,
                details: error.issues,
            },
            { status: 400 }
        );
    }
    
    if (error instanceof AppError) {
      return NextResponse.json(
        { 
          error: error.message,
          code: error.code,
          details: error.details 
        }, 
        { status: error.statusCode }
      )
    }
    
    return NextResponse.json(
      { 
        error: 'Failed to create compliance record',
        code: ErrorCode.INTERNAL_ERROR,
        requestId: requestId
      }, 
      { status: 500 }
    )
  }
}