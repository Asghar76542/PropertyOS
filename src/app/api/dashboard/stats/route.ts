import { NextResponse } from 'next/server'
import { db } from '@/lib/db'
import { getServerSession } from 'next-auth'
import { authOptions } from '@/lib/auth'
import { logger } from '@/lib/logger'
import { AppError, ErrorCode, generateRequestId } from '@/lib/errors'

export async function GET(request: Request) {
  const requestId = generateRequestId();
  let session;
  try {
    session = await getServerSession(authOptions)
    if (!session?.user?.id) {
      throw new AppError('Unauthorized', 401, ErrorCode.UNAUTHORIZED);
    }

    // Verify user has proper role
    if (session.user.role !== 'LANDLORD' && session.user.role !== 'ADMIN') {
      throw new AppError('Insufficient permissions', 403, ErrorCode.INSUFFICIENT_PERMISSIONS);
    }

    const landlordId = session.user.id;

    const [properties, payments, maintenanceRequests, complianceRecords] = await db.$transaction([
      db.property.findMany({
        where: { landlordId: landlordId },
        select: { id: true, monthlyRent: true, isAvailable: true }
      }),
      db.payment.findMany({
        where: { landlordId: landlordId },
        select: { amount: true, processingFee: true, status: true, paymentDate: true }
      }),
      db.maintenanceRequest.findMany({
        where: { property: { landlordId: landlordId } },
        select: { status: true }
      }),
      db.complianceRecord.findMany({
        where: { property: { landlordId: landlordId } },
        select: { status: true }
      })
    ]);

    const propertiesCount = properties.length;
    const totalMonthlyRent = properties.reduce((sum, property) => sum + property.monthlyRent, 0);
    const occupiedProperties = properties.filter(p => !p.isAvailable).length;
    const occupancyRate = propertiesCount > 0 ? (occupiedProperties / propertiesCount) * 100 : 0;
    const totalIncome = payments.filter(p => p.status === 'COMPLETED').reduce((sum, payment) => sum + (payment.amount - payment.processingFee), 0);

    const stats = {
      properties: {
        total: propertiesCount,
        occupied: occupiedProperties,
        vacant: propertiesCount - occupiedProperties,
        occupancyRate: Math.round(occupancyRate * 100) / 100
      },
      finances: {
        totalMonthlyRent: Math.round(totalMonthlyRent * 100) / 100,
        totalIncome: Math.round(totalIncome * 100) / 100,
        averageRent: propertiesCount > 0 ? Math.round((totalMonthlyRent / propertiesCount) * 100) / 100 : 0
      },
      maintenance: {
        open: maintenanceRequests.filter(r => r.status === 'OPEN').length,
        inProgress: maintenanceRequests.filter(r => r.status === 'IN_PROGRESS').length,
        completed: maintenanceRequests.filter(r => r.status === 'COMPLETED').length,
        total: maintenanceRequests.length
      },
      compliance: {
        compliant: complianceRecords.filter(r => r.status === 'COMPLIANT').length,
        pending: complianceRecords.filter(r => r.status === 'PENDING').length,
        overdue: complianceRecords.filter(r => r.status === 'OVERDUE').length,
        total: complianceRecords.length
      }
    };

    const propertyIds = properties.map(p => p.id);

    const [recentMaintenanceRequests, recentComplianceRecords, recentPayments] = await db.$transaction([
        db.maintenanceRequest.findMany({
            where: { propertyId: { in: propertyIds } },
            include: {
                property: {
                    select: { address: true, city: true }
                }
            },
            orderBy: { createdAt: 'desc' },
            take: 5
        }),
        db.complianceRecord.findMany({
            where: { propertyId: { in: propertyIds } },
            orderBy: { dueDate: 'asc' },
            take: 5
        }),
        db.payment.findMany({
            where: { landlordId: landlordId },
            include: {
                property: {
                    select: { address: true, city: true }
                }
            },
            orderBy: { createdAt: 'desc' },
            take: 5
        })
    ]);


    return NextResponse.json({
      stats,
      recentData: {
        maintenanceRequests: recentMaintenanceRequests,
        complianceRecords: recentComplianceRecords,
        recentPayments: recentPayments
      }
    })
  } catch (error) {
    logger.error('Error fetching dashboard stats:', { error, requestId, userId: session?.user?.id })
    
    if (error instanceof AppError) {
      return NextResponse.json(
        { 
          error: error.message,
          code: error.code,
          details: error.details 
        }, 
        { status: error.statusCode }
      )
    }
    
    return NextResponse.json(
      { 
        error: 'Failed to fetch dashboard statistics',
        code: ErrorCode.INTERNAL_ERROR,
        requestId: requestId
      }, 
      { status: 500 }
    )
  }
}