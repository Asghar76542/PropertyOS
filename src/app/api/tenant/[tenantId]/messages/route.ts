import { NextResponse } from 'next/server'
import { db } from '@/lib/db'
import { UserRole } from '@prisma/client'

export async function GET(
  request: Request,
  context: { params: Promise<{ tenantId: string }> }
) {
  try {
    const { tenantId } = await context.params

    // Verify the tenant exists
    const tenant = await db.user.findUnique({
      where: {
        id: tenantId,
        role: UserRole.TENANT
      }
    })

    if (!tenant) {
      return NextResponse.json({ error: 'Tenant not found' }, { status: 404 })
    }

    // Get messages for this tenant
    const messages = await db.message.findMany({
      where: {
        toId: tenantId
      },
      include: {
        from: {
          select: {
            name: true,
            role: true
          }
        }
      },
      orderBy: {
        createdAt: 'desc'
      },
      take: 20
    })

    // Count unread messages
    const unreadCount = await db.message.count({
      where: {
        toId: tenantId,
        read: false
      }
    })

    const formattedMessages = messages.map(message => ({
      id: message.id,
      from: `${message.from.name} (${message.from.role === 'LANDLORD' ? 'Landlord' : 'System'})`,
      subject: message.subject,
      message: message.body,
      date: message.createdAt.toISOString(),
      unread: !message.read
    }))

    return NextResponse.json({
      unreadMessages: unreadCount,
      recentMessages: formattedMessages
    })
  } catch (error) {
    console.error('Error fetching messages:', error)
    return NextResponse.json({ error: 'Internal Server Error' }, { status: 500 })
  }
}

export async function PATCH(
  request: Request,
  context: { params: Promise<{ tenantId: string }> }
) {
  try {
    const { tenantId } = await context.params
    const { messageId } = await request.json()

    // Mark message as read
    const updatedMessage = await db.message.update({
      where: {
        id: messageId,
        toId: tenantId
      },
      data: {
        read: true
      }
    })

    return NextResponse.json({ success: true })
  } catch (error) {
    console.error('Error marking message as read:', error)
    return NextResponse.json({ error: 'Internal Server Error' }, { status: 500 })
  }
}
