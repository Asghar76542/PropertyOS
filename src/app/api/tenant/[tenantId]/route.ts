import { NextRequest, NextResponse } from 'next/server'
import { db } from '@/lib/db'
import { UserRole, PaymentStatus, MaintenanceStatus } from '@prisma/client'
import { getServerSession } from 'next-auth'
import { authOptions } from '@/lib/auth'
import { logger } from '@/lib/logger'
import { AppError, ErrorCode, generateRequestId } from '@/lib/errors'

export async function GET(
  request: NextRequest,
  { params }: { params: { tenantId: string } }
) {
  const requestId = generateRequestId();
  let session;
  try {
    session = await getServerSession(authOptions);
    if (!session?.user?.id) {
      throw new AppError('Unauthorized', 401, ErrorCode.UNAUTHORIZED);
    }

    const { tenantId } = params;

    // RBAC: Allow access only to the tenant themselves or an ADMIN.
    // Landlord access would require checking ownership of the property/tenancy.
    if (session.user.role !== 'ADMIN' && session.user.id !== tenantId) {
        throw new AppError('Insufficient permissions', 403, ErrorCode.INSUFFICIENT_PERMISSIONS);
    }

    const tenant = await db.user.findUnique({
      where: {
        id: tenantId,
        role: UserRole.TENANT
      }
    });

    if (!tenant) {
      throw new AppError('Tenant not found', 404, ErrorCode.NOT_FOUND);
    }

    const tenancy = await db.tenancy.findFirst({
      where: { tenantId: tenantId },
      include: {
        property: {
          include: {
            landlord: true
          }
        }
      }
    });

    if (!tenancy) {
      throw new AppError('No tenancy found for this tenant', 404, ErrorCode.NOT_FOUND);
    }

    const { property } = tenancy;
    const { landlord } = property;

    const payments = await db.payment.findMany({
      where: { tenancyId: tenancy.id },
      orderBy: { dueDate: 'desc' },
      take: 10
    });

    const documents = await db.document.findMany({
      where: { tenancyId: tenancy.id },
      orderBy: { createdAt: 'desc' },
      take: 10
    });

    const maintenanceRequests = await db.maintenanceRequest.findMany({
      where: { propertyId: property.id },
      orderBy: { createdAt: 'desc' },
      take: 10
    });

    const messages = await db.message.findMany({
      where: { toId: tenantId },
      include: {
        from: { select: { name: true, role: true } }
      },
      orderBy: { createdAt: 'desc' },
      take: 5
    });

    const unreadCount = await db.message.count({
      where: { toId: tenantId, read: false }
    });

    const totalPaid = payments
      .filter(p => p.status === PaymentStatus.COMPLETED)
      .reduce((sum, payment) => sum + payment.amount, 0);

    const pendingPayments = payments.filter(p => p.status === PaymentStatus.PENDING);
    const balanceDue = pendingPayments.reduce((sum, payment) => sum + payment.amount, 0);
    const nextDueDate = pendingPayments.length > 0 
      ? pendingPayments[0].dueDate 
      : null;

    const openRequests = maintenanceRequests.filter(r => r.status === MaintenanceStatus.OPEN).length;
    const inProgressRequests = maintenanceRequests.filter(r => r.status === MaintenanceStatus.IN_PROGRESS).length;
    const completedRequests = maintenanceRequests.filter(r => r.status === MaintenanceStatus.COMPLETED).length;

    const data = {
      personalInfo: {
        name: tenant.name || '',
        email: tenant.email,
        moveInDate: tenancy.startDate.toISOString(),
        leaseEnd: tenancy.endDate.toISOString(),
        propertyAddress: property.address,
        propertyCity: property.city,
        landlordName: landlord.name || '',
        landlordEmail: landlord.email,
      },
      financials: {
        monthlyRent: tenancy.monthlyRent,
        securityDeposit: tenancy.deposit,
        balanceDue: balanceDue,
        lastPaymentDate: payments.find(p => p.status === PaymentStatus.COMPLETED)?.paymentDate?.toISOString() || null,
        nextPaymentDue: nextDueDate?.toISOString() || null,
        paymentHistory: payments.map(payment => ({
          id: payment.id,
          amount: payment.amount,
          date: payment.paymentDate?.toISOString() || payment.dueDate.toISOString(),
          status: payment.status,
          method: (payment as any).method || null // Added `any` cast until schema is updated
        }))
      },
      maintenance: {
        openRequests,
        inProgress: inProgressRequests,
        completed: completedRequests,
        recentRequests: maintenanceRequests.slice(0, 5).map(request => ({
          id: request.id,
          title: request.title,
          status: request.status,
          priority: request.priority,
          submittedDate: request.createdAt.toISOString(),
          estimatedCompletion: request.dueDate?.toISOString() || null
        }))
      },
      documents: {
        leaseAgreement: {
          available: documents.some(d => d.documentType === 'LEASE_AGREEMENT'),
          lastUpdated: tenancy.startDate.toISOString()
        },
        importantDocuments: documents.map(doc => ({
          id: doc.id,
          name: doc.originalName,
          type: doc.documentType,
          uploadDate: doc.createdAt.toISOString(),
          size: doc.fileSize
        }))
      },
      communications: {
        unreadMessages: unreadCount,
        recentMessages: messages.map(message => ({
          id: message.id,
          from: `${message.from.name} (${message.from.role === 'LANDLORD' ? 'Landlord' : 'System'})`,
          subject: message.subject,
          message: message.body,
          date: message.createdAt.toISOString(),
          unread: !message.read
        }))
      }
    };

    return NextResponse.json(data);

  } catch (error) {
    logger.error('[TENANT_API]', { error, requestId, userId: session?.user?.id });
    
    if (error instanceof AppError) {
      return NextResponse.json(
        { error: error.message, code: error.code, details: error.details }, 
        { status: error.statusCode }
      );
    }
    
    return NextResponse.json(
      { error: 'Internal Server Error', code: ErrorCode.INTERNAL_ERROR, requestId }, 
      { status: 500 }
    );
  }
}
