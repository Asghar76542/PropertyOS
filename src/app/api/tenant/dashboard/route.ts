import { NextResponse } from 'next/server'
import { db } from '@/lib/db'
import { UserRole } from '@prisma/client'
import { getServerSession } from 'next-auth'
import { authOptions } from '@/lib/auth'
import { logger } from '@/lib/logger'
import { AppError, ErrorCode, generateRequestId } from '@/lib/errors'

export async function GET(request: Request) {
  const requestId = generateRequestId();
  let session;
  try {
    session = await getServerSession(authOptions)
    if (!session?.user?.id) {
      throw new AppError('Unauthorized', 401, ErrorCode.UNAUTHORIZED);
    }

    if (session.user.role !== 'TENANT') {
      throw new AppError('Insufficient permissions', 403, ErrorCode.INSUFFICIENT_PERMISSIONS);
    }

    const tenantId = session.user.id;

    const tenant = await db.user.findUnique({
        where: { id: tenantId }
    });

    if (!tenant) {
      throw new AppError('Tenant not found', 404, ErrorCode.NOT_FOUND);
    }

    // Fetch tenancy information with all related data
    const tenancy = await db.tenancy.findFirst({
      where: { tenantId: tenant.id },
      include: {
        property: {
          include: {
            landlord: true,
          },
        },
        payments: {
          orderBy: {
            dueDate: 'desc',
          },
        },
        documents: {
          orderBy: {
            createdAt: 'desc',
          },
        },
      },
    })

    if (!tenancy) {
      throw new AppError('No tenancy found for the tenant', 404, ErrorCode.NOT_FOUND);
    }

    // Fetch maintenance requests for the property
    const maintenanceRequests = await db.maintenanceRequest.findMany({
      where: { propertyId: tenancy.propertyId },
      orderBy: {
        createdAt: 'desc',
      },
    })

    // Calculate financial data
    const completedPayments = tenancy.payments.filter(p => p.status === 'COMPLETED')
    const lastPayment = completedPayments.length > 0 ? completedPayments[0] : null
    const pendingPayments = tenancy.payments.filter(p => p.status === 'PENDING')
    const totalPending = pendingPayments.reduce((sum, p) => sum + p.amount, 0)

    // Calculate next payment due date (first pending payment or monthly cycle)
    const nextDueDate = pendingPayments.length > 0 
      ? pendingPayments[0].dueDate 
      : new Date(Date.now() + 30 * 24 * 60 * 60 * 1000) // 30 days from now

    const data = {
      personalInfo: {
        name: tenant.name || 'N/A',
        email: tenant.email,
        moveInDate: tenancy.startDate.toISOString(),
        leaseEnd: tenancy.endDate.toISOString(),
        propertyAddress: tenancy.property.address,
        propertyCity: tenancy.property.city,
        landlordName: tenancy.property.landlord.name || 'N/A',
        landlordEmail: tenancy.property.landlord.email,
        landlordId: tenancy.property.landlord.id,
      },
      financials: {
        monthlyRent: tenancy.monthlyRent,
        securityDeposit: tenancy.deposit,
        balanceDue: totalPending,
        lastPaymentDate: lastPayment?.paymentDate?.toISOString() || 'N/A',
        nextPaymentDue: nextDueDate.toISOString(),
        paymentHistory: tenancy.payments.map(p => ({
          id: p.id,
          amount: p.amount,
          date: p.paymentDate?.toISOString() || p.dueDate.toISOString(),
          status: p.status,
          method: p.method || null,
        })),
      },
      maintenance: {
        openRequests: maintenanceRequests.filter(r => r.status === 'OPEN').length,
        inProgress: maintenanceRequests.filter(r => r.status === 'IN_PROGRESS').length,
        completed: maintenanceRequests.filter(r => r.status === 'COMPLETED').length,
        recentRequests: maintenanceRequests.map(r => ({
          id: r.id,
          title: r.title,
          status: r.status,
          priority: r.priority,
          submittedDate: r.createdAt.toISOString(),
          estimatedCompletion: r.dueDate?.toISOString() || 'N/A',
        })),
      },
      documents: {
        leaseAgreement: {
          available: tenancy.documents.some(d => d.documentType === 'LEASE_AGREEMENT'),
          lastUpdated: tenancy.documents.find(d => d.documentType === 'LEASE_AGREEMENT')?.updatedAt.toISOString() || tenancy.startDate.toISOString(),
        },
        importantDocuments: tenancy.documents.map(d => ({
          id: d.id,
          name: d.originalName,
          type: d.documentType.replace(/_/g, ' ').toLowerCase().replace(/\b\w/g, l => l.toUpperCase()),
          uploadDate: d.createdAt.toISOString(),
          size: `${(d.fileSize / 1024).toFixed(0)} KB`,
        })),
      },
      communications: {
        unreadMessages: 0,
        recentMessages: []
      }
    };

    return NextResponse.json(data)
  } catch (error) {
    logger.error('[TENANT_DASHBOARD_API]', { error, requestId, userId: session?.user?.id })
    
    if (error instanceof AppError) {
      return NextResponse.json(
        { 
          error: error.message,
          code: error.code,
          details: error.details 
        }, 
        { status: error.statusCode }
      )
    }
    
    return NextResponse.json(
      { 
        error: 'Internal Server Error',
        code: ErrorCode.INTERNAL_ERROR,
        requestId: requestId
      }, 
      { status: 500 }
    )
  }
}