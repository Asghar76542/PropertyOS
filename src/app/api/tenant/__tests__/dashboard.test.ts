import { GET } from '../dashboard/route';
import { getServerSession } from 'next-auth';
import { db } from '@/lib/db';
import { logger } from '@/lib/logger';
import { NextRequest } from 'next/server';
import { UserRole } from '@prisma/client';

// Mock dependencies
jest.mock('next-auth');

// Manual mock for the Prisma client to ensure deep object structure
jest.mock('@/lib/db', () => ({
  __esModule: true,
  db: {
    user: {
      findUnique: jest.fn(),
    },
    tenancy: {
      findFirst: jest.fn(),
    },
    maintenanceRequest: {
      findMany: jest.fn(),
    },
  },
}));

jest.mock('@/lib/logger', () => ({
  logger: {
    error: jest.fn(),
  },
}));

// Typecast mocks for easier usage
const mockedGetServerSession = getServerSession as jest.Mock;
const mockedDb = db as jest.Mocked<typeof db>;

const MOCK_TENANT_ID = 'user-tenant-123';
const MOCK_LANDLORD_ID = 'user-landlord-456';
const MOCK_PROPERTY_ID = 'prop-123';
const MOCK_TENANCY_ID = 'tenancy-123';

describe('GET /api/tenant/dashboard', () => {
  beforeEach(() => {
    // Clear all mocks before each test
    jest.clearAllMocks();
  });

  it('should return the correct dashboard data shape for a valid tenant', async () => {
    // 1. Setup Mock Data and Session
    mockedGetServerSession.mockResolvedValue({
      user: {
        id: MOCK_TENANT_ID,
        role: UserRole.TENANT,
      },
    });

    (mockedDb.user.findUnique as jest.Mock).mockResolvedValue({
      id: MOCK_TENANT_ID,
      name: 'Test Tenant',
      email: 'tenant@test.com',
      role: UserRole.TENANT,
    } as any);

    (mockedDb.tenancy.findFirst as jest.Mock).mockResolvedValue({
      id: MOCK_TENANCY_ID,
      startDate: new Date('2024-01-01T00:00:00.000Z'),
      endDate: new Date('2025-01-01T00:00:00.000Z'),
      monthlyRent: 1200,
      deposit: 1500,
      propertyId: MOCK_PROPERTY_ID,
      property: {
        id: MOCK_PROPERTY_ID,
        address: '123 Fake St',
        city: 'Testville',
        landlord: {
          name: 'Test Landlord',
          email: 'landlord@test.com',
        },
      },
      payments: [
        {
          id: 'payment-1',
          amount: 1200,
          status: 'COMPLETED',
          dueDate: new Date('2024-08-01T00:00:00.000Z'),
          paymentDate: new Date('2024-08-01T00:00:00.000Z'),
          method: 'bank_transfer',
        },
        {
          id: 'payment-2',
          amount: 1200,
          status: 'PENDING',
          dueDate: new Date('2024-09-01T00:00:00.000Z'),
          paymentDate: null,
          method: null,
        },
      ],
      documents: [
        {
            id: 'doc-1',
            documentType: 'LEASE_AGREEMENT',
            originalName: 'lease.pdf',
            fileSize: 204800,
            createdAt: new Date('2024-01-01T00:00:00.000Z'),
            updatedAt: new Date('2024-01-01T00:00:00.000Z'),
        }
      ],
    } as any);

    (mockedDb.maintenanceRequest.findMany as jest.Mock).mockResolvedValue([
        {
            id: 'req-1',
            title: 'Leaky Faucet',
            status: 'OPEN',
            priority: 'MEDIUM',
            createdAt: new Date('2024-08-15T00:00:00.000Z'),
            dueDate: null,
        }
    ]);

    // 2. Call the API Route Handler
    const request = new NextRequest('http://localhost/api/tenant/dashboard');
    const response = await GET(request);
    const data = await response.json();

    // 3. Assertions
    expect(response.status).toBe(200);
    // Snapshot test to ensure the data structure remains consistent
    expect(data).toMatchSnapshot();
    // Check that no errors were logged
    expect(logger.error).not.toHaveBeenCalled();
  });

  it('should return a 401 Unauthorized error if there is no session', async () => {
    mockedGetServerSession.mockResolvedValue(null);

    const request = new NextRequest('http://localhost/api/tenant/dashboard');
    const response = await GET(request);
    const data = await response.json();

    expect(response.status).toBe(401);
    expect(data.code).toBe('UNAUTHORIZED');
    expect(logger.error).toHaveBeenCalled();
  });

    it('should return a 403 Insufficient Permissions error for a non-tenant user', async () => {
    mockedGetServerSession.mockResolvedValue({
        user: {
            id: MOCK_LANDLORD_ID,
            role: UserRole.LANDLORD,
        },
    });

    const request = new NextRequest('http://localhost/api/tenant/dashboard');
    const response = await GET(request);
    const data = await response.json();

    expect(response.status).toBe(403);
    expect(data.code).toBe('INSUFFICIENT_PERMISSIONS');
    expect(logger.error).toHaveBeenCalled();
  });
});