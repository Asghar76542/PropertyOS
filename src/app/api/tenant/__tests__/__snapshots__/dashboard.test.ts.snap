// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/tenant/dashboard should return the correct dashboard data shape for a valid tenant 1`] = `
{
  "communications": {
    "recentMessages": [],
    "unreadMessages": 0,
  },
  "documents": {
    "importantDocuments": [
      {
        "id": "doc-1",
        "name": "lease.pdf",
        "size": "200 KB",
        "type": "Lease Agreement",
        "uploadDate": "2024-01-01T00:00:00.000Z",
      },
    ],
    "leaseAgreement": {
      "available": true,
      "lastUpdated": "2024-01-01T00:00:00.000Z",
    },
  },
  "financials": {
    "balanceDue": 1200,
    "lastPaymentDate": "2024-08-01T00:00:00.000Z",
    "monthlyRent": 1200,
    "nextPaymentDue": "2024-09-01T00:00:00.000Z",
    "paymentHistory": [
      {
        "amount": 1200,
        "date": "2024-08-01T00:00:00.000Z",
        "id": "payment-1",
        "method": "bank_transfer",
        "status": "COMPLETED",
      },
      {
        "amount": 1200,
        "date": "2024-09-01T00:00:00.000Z",
        "id": "payment-2",
        "method": null,
        "status": "PENDING",
      },
    ],
    "securityDeposit": 1500,
  },
  "maintenance": {
    "completed": 0,
    "inProgress": 0,
    "openRequests": 1,
    "recentRequests": [
      {
        "estimatedCompletion": "N/A",
        "id": "req-1",
        "priority": "MEDIUM",
        "status": "OPEN",
        "submittedDate": "2024-08-15T00:00:00.000Z",
        "title": "Leaky Faucet",
      },
    ],
  },
  "personalInfo": {
    "email": "tenant@test.com",
    "landlordEmail": "landlord@test.com",
    "landlordName": "Test Landlord",
    "leaseEnd": "2025-01-01T00:00:00.000Z",
    "moveInDate": "2024-01-01T00:00:00.000Z",
    "name": "Test Tenant",
    "propertyAddress": "123 Fake St",
    "propertyCity": "Testville",
  },
}
`;
